src/lib/storage/variable_string_dictionary/variable_string_dictionary_encoder.hpp

```diff
@@ -0,0 +1,122 @@
+    auto string_to_chunk_offsets = std::unordered_map<pmr_string, std::vector<ChunkOffset>>();
+    auto segment_size = uint32_t{0};
+
+    // Iterate over segment, save all values and save to values the chunk_offset.
+    segment_iterable.with_iterators([&](auto segment_it, const auto segment_end) {
+      segment_size = std::distance(segment_it, segment_end);
+      dense_values.reserve(segment_size);  // Potentially overallocate for segments with NULLs.
+      null_values.resize(segment_size);    // Resized to size of segment.
+
+      for (auto current_position = size_t{0}; segment_it != segment_end; ++segment_it, ++current_position) {
+        const auto segment_item = *segment_it;
+        if (!segment_item.is_null()) {
+          const auto& segment_value = segment_item.value();
+          dense_values.push_back(segment_value);
+          string_to_chunk_offsets[segment_value].push_back(ChunkOffset(current_position));
+        } else {
+          null_values[current_position] = true;
+        }
+      }
+    });
+
+    // Eliminate duplicate strings.
+    std::sort(dense_values.begin(), dense_values.end());
+    dense_values.erase(std::unique(dense_values.begin(), dense_values.end()), dense_values.cend());
+    dense_values.shrink_to_fit();
+
+    // Compute total compressed data size.
+    auto total_size = uint32_t{0};
+    for (const auto& value : dense_values) {
+      total_size += value.size() + 1;
+    }
+
+    // uniform character array containing all distinct strings
+    auto klotz = std::make_shared<pmr_vector<char>>(pmr_vector<char>(total_size));
+    // We assume segment size up to 4 GByte.
+    // Maps string to offset in klotz.
+    auto string_offsets = std::unordered_map<pmr_string, uint32_t>();
+    // Maps string to ValueID for attribute vector.
+    auto string_value_ids = std::unordered_map<pmr_string, ValueID>();
+    auto current_offset = uint32_t{0};
+    auto current_value_id = ValueID{0};
+
+    // Construct klotz.
+    for (const auto& value : dense_values) {
+      memcpy(klotz->data() + current_offset, value.c_str(), value.size());
+      string_offsets[value] = current_offset;
+      string_value_ids[value] = current_value_id++;
+      current_offset += value.size() + 1;
+    }
+
+    // Maps ChunkOffset to ValueID.
+    auto chunk_offset_to_value_id = std::make_shared<pmr_vector<uint32_t>>(segment_size);
+    auto offset_vector = std::make_shared<pmr_vector<uint32_t>>(pmr_vector<uint32_t>(dense_values.size()));
+    offset_vector->shrink_to_fit();
+
+    // Construct attribute and offset vector.
+    for (const auto& [string, chunk_offsets] : string_to_chunk_offsets) {
+      const auto here_offset = string_offsets[string];
+      const auto here_value_id = string_value_ids[string];
+      (*offset_vector)[here_value_id] = here_offset;
+      for (const auto chunk_offset : chunk_offsets) {
+        (*chunk_offset_to_value_id)[chunk_offset] = here_value_id;
+      }
```

Comment:
You can think about using a more descriptive variable name here and also, `here_offset` is only used once, so I'd suggest using something like this:

```suggestion
      const auto value_id = string_value_ids[string];
      (*offset_vector)[value_id] = string_offsets[string];
      for (const auto chunk_offset : chunk_offsets) {
        (*chunk_offset_to_value_id)[chunk_offset] = value_id;
      }
```

Reply:
We chose to use variables here to allow for easier debugging. During development there were some issues with evaluation of expressions. We did change the names to `value_id` and `offset`.